### PR TITLE
RCL - aka Razor UI Class

### LINKED ISSUE #5409: UI as a library - Package Razor pages and views as reusable libraries
aspnet/Razor#1809

### LINKED ISSUE #6073: Add a doc for creating and using a Razor Class Library(RCL) project
The topic should describe the following:
- How to find the RCL project in VS dialog (net461 & netcoreapp2.1 TFM's)
- Creating a new project and referencing it in a newly created Razor Pages(RP) web application project
- Launch the RP project and access the RCL project's page
- Override the RCL project's page in RP project and launch again to see the new changes.
- Table of Contents should probably have something saying "Creating a Razor Class Library project" or "Creating reusable UI using Razor Class Library project"

cc @danroth27 @mkArtakMSFT @Rick-Anderson 

@rick-anderson edit below:
Dup of #5409
aspnet/Razor#1809

### PR DESCRIPTION
**Pull request by Rick-Anderson**
Fixes #5409 and Fixes #6073
[Review URL](https://review.docs.microsoft.com/en-us/aspnet/core/mvc/razor-pages/ui-class?view=aspnetcore-1.0&branch=pr-en-us-5980)
See https://github.com/aspnet/Razor/issues/1809
See notes in #5790 
`@javiercn and @pranavkm` are the right people to review this and ping `@kichalla` on RCL

### CODE REVIEW

**Comment by Rick-Anderson** [comment]
@javiercn or  @pranavkm - please review. [Review URL](https://review.docs.microsoft.com/en-us/aspnet/core/mvc/razor-pages/ui-class?view=aspnetcore-1.0&branch=pr-en-us-5980)

**Comment by Rick-Anderson** [comment]
@javiercn or @pranavkm - hold off while I add the RCL template.

**Comment by Rick-Anderson** [comment]
@kichalla,  @javiercn or @pranavkm This is ready for review. You guys can flip coins on who reviews. Next week is fine.

@kichalla the fwlink should point to `https://docs.microsoft.com/aspnet/core/mvc/razor-pages/ui-class` once this is merged to live.

**Comment by Rick-Anderson** [comment]
@pranavkm  should be ready for review.

**Comment by Rick-Anderson** [comment]
@pranavkm  should be ready for review.

File: aspnetcore/mvc/razor-pages/ui-class.md [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+```
+
+For more information, see [dotnet new](/dotnet/core/tools/dotnet-new).
+
+------
+Add Razor files to the RCL.
+
+## Razor Class Library restrictions.
+
+* Razor files must be under the *Areas* folder.
```
Comment (pranavkm) [inline aspnetcore/mvc/razor-pages/ui-class.md]: This isn't a requirement, we just think using Areas is a better model for sharing isolated pieces of UI libraries.

File: aspnetcore/mvc/razor-pages/ui-class.md [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+* The ASP.NET Core runtime does not search for partial files in the RCL */Areas/{FeatureName}/Pages/Shared/* folder.
```
Comment (pranavkm) [inline aspnetcore/mvc/razor-pages/ui-class.md]: Did you see this when you were testing? This should work just fine

File: aspnetcore/mvc/razor-pages/ui-class.md [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+The RCL can be referenced by:
+
+* NuGet package. See [Creating NuGet packages](/nuget/create-packages/creating-a-package) and [dotnet add package](/dotnet/core/tools/dotnet-add-package) and [Create and publish a NuGet package](/nuget/quickstart/create-and-publish-a-package-using-visual-studio).
+* DLLs - for example, *{ProjectName}.dll* and *{ProjectName}.Views.dll*. See [dotnet-add reference](/dotnet/core/tools/dotnet-add-reference). *{ProjectName}.Views.dll* contains the compiled Razor content.
```
Comment (pranavkm) [inline aspnetcore/mvc/razor-pages/ui-class.md]: Could we say project references here instead? In general, we don't want to recommend referencing loose binaries as a way to share things.

File: aspnetcore/mvc/razor-pages/ui-class.md [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+* Add a Razor partial view file named *RazorUIClassLib/Areas/MyFeature/Pages/Shared/_Message.cshtml*.
+* Replace the markup of *RazorUIClassLib/Areas/MyFeature/Pages/Shared/_Message.cshtml* with the following code:
+
+[!code-html[Main](ui-class/samples/cli/RazorUIClassLib/Areas/MyFeature/Pages/Shared/_Message.cshtml)]
+
+* Copy the *_ViewStart.cshtml* file from the WebApp1 project to  *RazorUIClassLib/Areas/MyFeature/Pages/_ViewStart.cshtml*.
+
+  The [viewstart](xref:mvc/views/layout#running-code-before-each-view) file is required to use the layout of the Razor Pages project.
+
+* Replace the markup in *RazorUIClassLib/Areas/MyFeature/Pages/Shared/_Message.cshtml* with the following code:
```
Comment (pranavkm) [inline aspnetcore/mvc/razor-pages/ui-class.md]: This is repeating line 113.

File: aspnetcore/mvc/razor-pages/ui-class/samples/cli/RazorUIClassLib/RazorUIClassLib.csproj [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.AspNetCore.Mvc" Version="2.1.0-rc1-30661" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <Content Update="Areas\MyFeature\Pages\Shared\_Message.cshtml">
```
Comment (pranavkm) [inline aspnetcore/mvc/razor-pages/ui-class/samples/cli/RazorUIClassLib/RazorUIClassLib.csproj]: Remove. This is an annoying VS bug

File: aspnetcore/mvc/razor-pages/ui-class/samples/cli/RazorUIClassLib/RazorUIClassLib.csproj [hunk 6]
```diff
@@ -0,0 +1,16 @@
+<Project Sdk="Microsoft.NET.Sdk.Razor">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.AspNetCore.Mvc" Version="2.1.0-rc1-30661" />
```
Comment (pranavkm) [inline aspnetcore/mvc/razor-pages/ui-class/samples/cli/RazorUIClassLib/RazorUIClassLib.csproj]: `2.1.0-rc1-final` or just `2.1.0`. Think we're close to getting RTM bits any day now

File: aspnetcore/mvc/razor-pages/ui-class/samples/cli/WebApp1/WebApp1.csproj [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  </ItemGroup>
+
+
+  <ItemGroup>
+    <ProjectReference Include="..\RazorUIClassLib\RazorUIClassLib.csproj" />
+  </ItemGroup>
+
+
+  <ItemGroup>
+    <Content Update="Areas\MyFeature2\Pages\Index.cshtml">
```
Comment (pranavkm) [inline aspnetcore/mvc/razor-pages/ui-class/samples/cli/WebApp1/WebApp1.csproj]: Remove

File: aspnetcore/mvc/razor-pages/ui-class.md [hunk 8]
```diff
@@ -0,0 +1,217 @@
+---
+title: Reusable Razor UI in class libraries with ASP.NET Core
+author: Rick-Anderson
+description: Explains how to create reusable Razor Pages UI in a class library.
```
Comment (pranavkm) [inline aspnetcore/mvc/razor-pages/ui-class.md]: create reusable Razor ~~Pages~~ UI in a class library

**Pull request merged**